numpy/lib/_function_base_impl.py

```diff
@@ -538,9 +548,19 @@ def average(a, axis=None, weights=None, returned=False, *,
                     "Length of weights not compatible with specified axis.")
 
             # setup wgt to broadcast along axis
-            wgt = np.broadcast_to(wgt, (a.ndim-1)*(1,) + wgt.shape)
+            if ignore_nan:
+                wgt = np.broadcast_to(wgt, a.shape)
+            else:
+                wgt = np.broadcast_to(wgt, (a.ndim-1)*(1,) + wgt.shape)
             wgt = wgt.swapaxes(-1, axis)
 
+        if ignore_nan:
+            if not wgt.flags.writeable:
```

Comment:
why we suddenly care about this and all the previous operations modifying `wgt` ignored it?

Reply:
BCing makes wgt read-only, which stops _copyto from replacing nans with 0 in the following step.